DOC  improve documentation of copy=False in preprocessing functions

konstantinos-p opened the pull request:
#### Reference Issues/PRs

Example: Fixes #27307 (the issue was taken but had stalled)


#### What does this implement/fix? Explain your changes.

minmax_scale has an unexpected behavior when it is called with `X of dtype other than float` and `copy=False`. Specifically, even though `copy=False`, X is always first passed through the `~sklearn.utils.validation.check_array` function. If the dtype of X is not float check_array casts it to float triggering a copy. Any subsequent changes on X happen on the copy and not in-place.

I propose to first change the docstring of minmax_scale to reflect this unexpected behaviour. I also propose a check based on two calls of check_array that outputs a warning when this behaviour is detected.

#### Any other comments?

Other functions in the _data.py file probably have the same behavior and need to be modified accordingly. I'd be happy to change these as well.

konstantinos-p committed:
Change the docstring of minmax_scale + add warning when X dtype not float and copy=False

konstantinos-p commented:
@lesteve happy to have your thoughts!

konstantinos-p committed:
Fix errors raised by black and ruff

konstantinos-p committed:
Remove proposed warning, fix minmax_scale docstring according to StandardScaler docstring

lesteve commented on sklearn/preprocessing/_data.py:
```diff
@@ -649,6 +651,7 @@ def minmax_scale(X, feature_range=(0, 1), *, axis=0, copy=True):
     X = check_array(
         X, copy=False, ensure_2d=False, dtype=FLOAT_DTYPES, force_all_finite="allow-nan"
     )
+
```
Can you remove the added newline here, in order to avoid adding unrelated change?

lesteve commented:
> Once you are ok with the text I can add it to these functions as well, with some minor modifications.

I think your changes to the docstring look fine. I would wait for a second opinion before updating it in other places.

konstantinos-p committed:
Remove spurious newline in minmax_scale

lesteve commented:
Can you put it out of draft mode?

The CI error is unrelated.

betatim commented on sklearn/preprocessing/_data.py:
```diff
@@ -613,8 +613,10 @@ def minmax_scale(X, feature_range=(0, 1), *, axis=0, copy=True):
     copy : bool, default=True
-        Set to False to perform inplace scaling and avoid a copy (if the input
-        is already a numpy array).
+        If False, try to avoid a copy and do inplace scaling instead.
```
Some wording tweaks, not sure if they are significantly better. I am pretty sure though that it should be "in place", not inplace
```suggestion
        If False, try to avoid a copy and perform scaling in place.
```
```suggestion
        If False, try to avoid a copy and scale in place.
```

betatim commented:
Some wording tweak suggestions, otherwise the wording looks good and 👍 for applying it to all the similar locations you found.

konstantinos-p committed:
:art: Some wording tweaks sklearn/preprocessing/_data.py

Co-authored-by: Tim Head <betatim@gmail.com>

konstantinos-p committed:
Apply corrected text to scale, maxabs_scale, robust_scale, normalize, binarize, quantile_transform, power_transform

konstantinos-p committed:
Change inplace in the second sentence to in place to match the phrasing of the first sentence

geronimos commented:
I've seen the related issue today because I also got confused about the Transformer's behavior while setting `copy=False`. I much appreciate this PR.

I would like to suggest being more specific about the data types restriction by saying 
> NumPy array of dtype float

According to the note 
> e.g. if the data is not a NumPy array, is scipy.sparse CSR matrix, or is not of dtype float, a copy may still be returned.

I'd conclude that the data should either be a NumPy array **or** of dtype float. However, a NumPy array of dtype `int` is also copied. I assume the condition is NumPy array **and** dtype float.

I relate to @lesteve [comment in the issue](https://github.com/scikit-learn/scikit-learn/issues/27307#issuecomment-1708465143):

> `copy=False` only works if the input array dtype is a float dtype, i.e. `float64`, `float32` or `float16` right now. I guess maybe the documentation could be improved to mention this?
> 
> In your case the input array dtype is an int dtype.

@konstantinos-p @lesteve Is naming the restriction to dtypes applicable in your opinion?

lesteve commented on sklearn/preprocessing/_data.py:
```diff
@@ -2059,9 +2066,10 @@ def binarize(X, *, threshold=0.0, copy=True):
-        Set to False to perform inplace binarization and avoid a copy
-        (if the input is already a numpy array or a scipy.sparse CSR / CSC
-        matrix and if axis is 1).
+        If False, try to avoid a copy and scale in place.
```
I feel in this case (and maybe others) "scale" is not appropriate, so maybe stay closer to the original wording.

lesteve commented on sklearn/preprocessing/_data.py:
```diff
@@ -2059,9 +2066,10 @@ def binarize(X, *, threshold=0.0, copy=True):
-        (if the input is already a numpy array or a scipy.sparse CSR / CSC
-        matrix and if axis is 1).
+        If False, try to avoid a copy and scale in place.
+        This is not guaranteed to always work in place; e.g. if the data is
```
I think in this case (and maybe others) the previous wording is accurate that scipy CSR or CSC matrices are not copied, so you can not reuse the same doc here.

In general, the fact that a copy is made mostly depends on the `check_array` arguments, but the code following it may also need to be looked at closer.

lesteve replied:
I still think the previous docstring was accurate whether the new one is slightly misleading.

The code below is used, which means that it accepts CSR/CSC sparse matrix without copying it:
```py
X = check_array(X, accept_sparse=["csr", "csc"], copy=copy)
```

konstantinos-p committed:
Adapt the phrase scale, to more accurately reflect the transformation performed by each function

konstantinos-p committed:
Change the example when a copy occurs for the binarize function, fix other small typos

konstantinos-p commented:
Hello! I've made two commits. In the first, I changed the language for the binarize, quantile_transform, power_transform functions from "scale" to something that more accurately reflects what each function is doing. In the second, I fixed the examples that I give for when a copy occurs. I've checked that they are accurate. binarize returns a copy when it is called on something that is not a NumPy array. All the other functions return a copy "if the data is not a NumPy array, is scipy.sparse CSR matrix, or is not of dtype float".

lesteve commented:
FYI I have changed the title of your PR to remove the WIP and make it slighly more compact. This will also affect the commit message when we merge this.

lesteve commented on sklearn/preprocessing/_data.py:
```diff
@@ -1336,8 +1339,10 @@ def maxabs_scale(X, *, axis=0, copy=True):
+        If False, try to avoid a copy and scale in place.
+        This is not guaranteed to always work in place; e.g. if the data is
+        not a NumPy array, or is a scipy.sparse CSR matrix, or is not of dtype
+        float, a copy may still be returned.
```
`check_array(..., accept_sparse=("csr", "csc"), dtype=FLOAT_DTYPES)` means there is no copy if the matrix is a CSC or CSR matrix and the dtype is of kind float

lesteve commented on sklearn/preprocessing/_data.py:
```diff
@@ -1713,9 +1718,10 @@ def robust_scale(
+        If False, try to avoid a copy and scale in place.
+        This is not guaranteed to always work in place; e.g. if the data is
+        not a NumPy array, or is a scipy.sparse CSR matrix, or is not of dtype
+        float, a copy may still be returned.
```
Honestly I am started to think that aiming to be fully accurate in the docstring about the copy=False behaviour is not going to be workable. I would instead give 1 or 2 simple examples of case where data is copied even with `copy=False`.

For example in this case I would say something like "e.g. if the data is a Numpy array with a dtype not float, a copy will be returned". Basically let's only look at the first check_array call and give an example where even with `copy=False` it will return a copy.

In this particular case the first `check_array` call accept csr/csc so no copy but then it depends on the `RobustScaler`, for example the CSR will be turned into CSC in `RobustScaler.fit` which may share some indices (so partial copy?) ...

konstantinos-p replied:
There are two ways of formulating the docstring. 1) We present the cases where `copy = False` works exhaustively 2) we present examples where `copy=False` fails. I completely agree that the second approach is more feasible. 

As you mention, copies probably happen in different functions called by (in this case) maxabs_scale and simply looking at the check_array arguments is misleading. Is however the source of the copy relevant for the docstring? 

I see two ways moving forward. 

1) I create some sort of test that confirms copies are made in case `x`. For example, I've created a colab where I already check that a copy is made in all the cases that I've listed (even if these don't happen because of check_array, I'm not tracking the source). I can adapt it based on a proposed template. 

2) I include in all dosctrings only the example `data is not a NumPy array` which triggers a copy through `check_array` in all the functions that are part of this PR. (This is basically also what you are proposing if I understand correctly)

lesteve replied:
It feels like we are on the same line, i.e. giving one or two simple counter-examples, where copy=False still creates a copy so 2).

About ways for moving forward, I would say the testing approach seems too much work and not worth it.

So I would say:
- use the simple example of the numpy array with a non-float dtype everywhere we can, i.e. the first `check_array` call has `dtype=FLOAT_DTYPES`. This was the confusion in the original issue. More generally I am finding only now (sorry for realising this late) that "is not a Numpy array" is a bit unclear because a scipy sparse matrix is not a numpy array and yet in some cases scipy sparse matrix will not be copied.
- treat the remaining cases with more care (I don't have a good view how many there are)

> As you mention, copies probably happen in different functions called by (in this case) maxabs_scale and simply looking at the check_array arguments is misleading

Just for completeness, I was trying to say that looking at the first `check_array` is enough to find a counter-example where copy=False still creates a copy.

lesteve replied:
Also I would use the wording and use "a copy will be returned" instead of "a copy may be returned" for the counter-example we give.

In some way it would be nice to find a way to convey that this is a example where copy=False still returns a copy but that there may be (plenty of) others.

lesteve commented on sklearn/preprocessing/_data.py:
```diff
@@ -1826,9 +1832,10 @@ def normalize(X, norm="l2", *, axis=1, copy=True, return_norm=False):
+        If False, try to avoid a copy and normalize in place.
+        This is not guaranteed to always work in place; e.g. if the data is
+        not a NumPy array, or is a scipy.sparse CSR matrix, or is not of dtype
+        float, a copy may still be returned.
```
I think this needs to be tweaked

konstantinos-p committed:
Streamline text across all functions with single counterexample using wrong dtypes

lesteve committed:
lint

lesteve commented:
LGTM, I set auto-merge on this PR
